File: backend/app/schemas/stock.py
from pydantic import BaseModel, Field
from typing import Optional, Dict, Any, List
from datetime import datetime

class StockInfoResponse(BaseModel):
    """株式基本情報のレスポンススキーマ"""
    symbol: str
    name: str
    sector: Optional[str] = None
    industry: Optional[str] = None
    market_cap: Optional[float] = None
    current_price: Optional[float] = None
    pe_ratio: Optional[float] = None
    pb_ratio: Optional[float] = None
    peg_ratio: Optional[float] = None
    dividend_yield: Optional[float] = None
    beta: Optional[float] = None
    roe: Optional[float] = None
    roa: Optional[float] = None
    debt_to_equity: Optional[float] = None
    current_ratio: Optional[float] = None
    quick_ratio: Optional[float] = None
    gross_margin: Optional[float] = None
    operating_margin: Optional[float] = None
    profit_margin: Optional[float] = None
    revenue_growth: Optional[float] = None
    earnings_growth: Optional[float] = None
    fifty_two_week_high: Optional[float] = None
    fifty_two_week_low: Optional[float] = None
    volume: Optional[int] = None
    average_volume: Optional[int] = None
    shares_outstanding: Optional[int] = None
    float_shares: Optional[int] = None
    last_updated: str

class FinancialDataResponse(BaseModel):
    """財務データのレスポンススキーマ"""
    symbol: str
    financials: Dict[str, Any]
    balance_sheet: Dict[str, Any]
    cashflow: Dict[str, Any]
    last_updated: str

class HistoricalDataResponse(BaseModel):
    """履歴データのレスポンススキーマ"""
    symbol: str
    period: str
    data: List[Dict[str, Any]]
    last_updated: str

class FinancialScoreResponse(BaseModel):
    """財務スコアのレスポンススキーマ"""
    symbol: str
    overall_score: float = Field(..., ge=0, le=10, description="総合スコア (0-10)")
    detailed_scores: Dict[str, float]
    last_updated: str

class StockRequest(BaseModel):
    """株式情報取得リクエストスキーマ"""
    symbol: str = Field(..., min_length=1, max_length=10, description="株式ティッカーシンボル")

class HistoricalDataRequest(BaseModel):
    """履歴データ取得リクエストスキーマ"""
    symbol: str = Field(..., min_length=1, max_length=10, description="株式ティッカーシンボル")
    period: str = Field(default="1y", description="取得期間 (1d, 5d, 1mo, 3mo, 6mo, 1y, 2y, 5y, 10y, ytd, max)")

class ScreeningRequest(BaseModel):
    """スクリーニングリクエストスキーマ"""
    symbols: List[str] = Field(..., min_items=1, max_items=50, description="株式ティッカーシンボルのリスト")
    min_market_cap: Optional[float] = Field(None, ge=0, description="最小時価総額")
    max_pe_ratio: Optional[float] = Field(None, ge=0, description="最大PER")
    min_roe: Optional[float] = Field(None, ge=0, description="最小ROE")
    max_debt_to_equity: Optional[float] = Field(None, ge=0, description="最大負債比率")
    min_current_ratio: Optional[float] = Field(None, ge=0, description="最小流動比率")

class ScreeningResult(BaseModel):
    """スクリーニング結果のアイテム"""
    symbol: str
    name: str
    score: float
    market_cap: Optional[float] = None
    pe_ratio: Optional[float] = None
    roe: Optional[float] = None
    debt_to_equity: Optional[float] = None
    current_ratio: Optional[float] = None
    meets_criteria: bool

class ScreeningResponse(BaseModel):
    """スクリーニングレスポンススキーマ"""
    request_id: str
    total_symbols: int
    passed_symbols: int
    results: List[ScreeningResult]
    execution_time: float
    last_updated: str

class ErrorResponse(BaseModel):
    """エラーレスポンススキーマ"""
    error: str
    detail: Optional[str] = None
    timestamp: str

File: frontend/src/types/stock.ts
// 株式基本情報の型定義
export interface StockInfo {
  symbol: string;
  name: string;
  sector?: string;
  industry?: string;
  market_cap?: number;
  current_price?: number;
  pe_ratio?: number;
  pb_ratio?: number;
  peg_ratio?: number;
  dividend_yield?: number;
  beta?: number;
  roe?: number;
  roa?: number;
  debt_to_equity?: number;
  current_ratio?: number;
  quick_ratio?: number;
  gross_margin?: number;
  operating_margin?: number;
  profit_margin?: number;
  revenue_growth?: number;
  earnings_growth?: number;
  fifty_two_week_high?: number;
  fifty_two_week_low?: number;
  volume?: number;
  average_volume?: number;
  shares_outstanding?: number;
  float_shares?: number;
  last_updated: string;
}

// 財務データの型定義
export interface FinancialData {
  symbol: string;
  financials: Record<string, any>;
  balance_sheet: Record<string, any>;
  cashflow: Record<string, any>;
  last_updated: string;
}

// 履歴データの型定義
export interface HistoricalData {
  symbol: string;
  period: string;
  data: Array<{
    Date: string;
    Open: number;
    High: number;
    Low: number;
    Close: number;
    Volume: number;
  }>;
  last_updated: string;
}

// 財務スコアの型定義
export interface FinancialScore {
  symbol: string;
  overall_score: number;
  detailed_scores: {
    debt_score?: number;
    roe_score?: number;
    liquidity_score?: number;
    pe_score?: number;
    profit_score?: number;
  };
  last_updated: string;
}

// スクリーニングリクエストの型定義
export interface ScreeningRequest {
  symbols: string[];
  min_market_cap?: number;
  max_pe_ratio?: number;
  min_roe?: number;
  max_debt_to_equity?: number;
  min_current_ratio?: number;
}

// スクリーニング結果の型定義
export interface ScreeningResult {
  symbol: string;
  name: string;
  score: number;
  market_cap?: number;
  pe_ratio?: number;
  roe?: number;
  debt_to_equity?: number;
  current_ratio?: number;
  meets_criteria: boolean;
}

// スクリーニングレスポンスの型定義
export interface ScreeningResponse {
  request_id: string;
  total_symbols: number;
  passed_symbols: number;
  results: ScreeningResult[];
  execution_time: number;
  last_updated: string;
}

// API エラーレスポンスの型定義
export interface ApiError {
  error: string;
  detail?: string;
  timestamp: string;
}

// チャートデータの型定義
export interface ChartData {
  labels: string[];
  datasets: Array<{
    label: string;
    data: number[];
    borderColor: string;
    backgroundColor: string;
    fill: boolean;
  }>;
}

// ウォッチリストの型定義
export interface WatchListItem {
  id: number;
  symbol: string;
  name: string;
  current_price?: number;
  change?: number;
  change_percent?: number;
  target_price?: number;
  notes?: string;
  added_at: string;
}

// 検索結果の型定義
export interface SearchResult {
  symbol: string;
  name: string;
  relevance: number;
}

// APIレスポンスの汎用型
export interface ApiResponse<T> {
  data: T;
  success: boolean;
  message?: string;
}

// フィルター条件の型定義
export interface FilterConditions {
  minMarketCap?: number;
  maxPE?: number;
  minROE?: number;
  maxDebtToEquity?: number;
  minCurrentRatio?: number;
  sectors?: string[];
}

// ソート条件の型定義
export interface SortConditions {
  field: keyof StockInfo;
  direction: 'asc' | 'desc';
}

// ページネーションの型定義
export interface Pagination {
  page: number;
  limit: number;
  total: number;
  total_pages: number;
}